.github/workflows/main.yml

```diff
@@ -35,18 +36,64 @@ jobs:
 
       - name: Save docker image
         run: |
-          docker save research-template | gzip > /tmp/research-template.tar.gz
+          docker save research-template | pigz --fast > /tmp/research-template.tar.gz
 
       - name: Upload docker image
         uses: actions/upload-artifact@v4
         with:
             name: research-template-image
             path: /tmp/research-template.tar.gz
+            # Disable compression; the file is already compressed
+            compression-level: 0
 
-  publish:
+  test-devcontainer:
     needs: [build-and-test]
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Download Docker image
+      uses: actions/download-artifact@v4
+      with:
+        name: research-template-image
+        path: /tmp/image
+
+    - name: Import Docker image
+      run: docker load --input /tmp/image/research-template.tar.gz
+
+    - name: Tag Docker image for use with dev container
+      run: docker tag "$IMAGE_NAME" "$PUBLIC_IMAGE_NAME:$IMAGE_VERSION"
+
+    - name: Checkout research template temporary devcontainer dev branch
+      uses: actions/checkout@v4
+      with:
+        repository: opensafely/research-template
 
-    runs-on: ubuntu-22.04
+    - name: Checkout research-template-docker repository in subdirectory
+      uses: actions/checkout@v4
+      with:
+        path: 'research-template/research-template-docker'
+
+    - name: Install demjson package for jsonlint
+      # Necessary because jq doesn't yet have an option to strip comments.
+      # See https://github.com/jqlang/jq/issues/1571
+      run: |
+        sudo apt-get update
+        sudo apt-get install python3-demjson
+
+    - name: Check that we have correct version of image in devcontainer.json
+      run: |
+        devcontainer_image=$(jsonlint --allow comments --format-compactly ".devcontainer/devcontainer.json" | jq --raw-output .image)
+        test "$devcontainer_image" = "$PUBLIC_IMAGE_NAME:$IMAGE_VERSION"
```

Comment:
This is really two questions in one, I think. Let's separate them.

Reply:
# Use of `jq`

> Why don't we `grep` for `ghcr.io/opensafely-core/research-template:v0`? Then we need not worry about `jq` and `jsonlint`.

I actually used `grep` initially for this check.

I switched to `jq` to give a better assurance that we're actually validating the value of the `image` key in `devcontainer.json`.

And not accidentally matching something else. For example, comments that are allowed in Microsoft's JSON-with-comments variant that the `devcontainer.json` uses. (There are both single-line and multi-line comments allowed.)

Whether that extra robustness is warranted, I don't know.

It would be better to not have to use `jsonlint`, I agree.

Reply:
# Checking the `devcontainer.json` image version

> As I understand it, we're testing that the version of the image in `devcontainer.json` matches the version of the image we're building.

Yes. This is to ensure that we correctly build a devcontainer using the locally built image intended for test. And that we don't unwittingly test with an already published image; which is easy to do, because there's no way — as far as I know — of telling Docker to not pull remote images.

(Incidentally, maybe redirecting `ghcr.io` to `localhost` in `/etc/hosts` could also or additionally work here as a safeguard?)

However, we could also avoid this if we set both the local image, and modified the `devcontainer.json` to use `dev` as a version, I suppose, so maybe that's the better way here.

> If the test fails, then we should update `devcontainer.json` in opensafely/research-template. Consequently, isn't this a test of opensafely/research-template?

Kind of, yes.

It seemed initially reasonable to check this. But, let's consider upgrading to `v1`. To make this check pass and build a `v1` image, you'd actually need to upgrade the image version in the `research-template` `devcontainer.json` to a non-existent `v1` image first, which is the wrong way round.

So maybe it's fine for that to be just a manual upgrade chore, or something that Dependabot might eventually do. I think it's OK for us to be building a new image here ahead of us introducing it to the research-template.

(Dependabot can create PRs to upgrade `features` in `devcontainer.json`; not sure about Dependabot updating the `devcontainer.json` value for `image`.)

Reply:
To summarise,

* I've left `jsonlint` and `jq` in as I'm now modifying the `devcontainer.json`
* I've separated out tests that are more image focused — checking the packages — from checking the research-template runs in a dev container

Reply:
Now removed `jsonlint` and `jq` entirely as they're no longer necessary.